{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bb33250a-ec75-453a-9784-f2e05997911d",
   "metadata": {
    "collapsed": true,
    "editable": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package cellxgene_census:\n",
      "\n",
      "NAME\n",
      "    cellxgene_census - An API to facilitate use of the CZI Science CELLxGENE Census. The Census is a versioned container of single-cell data hosted at `CELLxGENE Discover`_.\n",
      "\n",
      "DESCRIPTION\n",
      "    The API is built on the `tiledbsoma` SOMA API, and provides a number of helper functions including:\n",
      "\n",
      "        * Open a named version of the Census, for use with the SOMA API\n",
      "        * Get a list of available Census versions, and for each version, a description\n",
      "        * Get a slice of the Census as an AnnData, for use with ScanPy\n",
      "        * Get the URI for, or directly download, underlying data in H5AD format\n",
      "\n",
      "    For more information on the API, visit the `cellxgene_census repo`_. For more information on SOMA, see the `tiledbsoma repo`_.\n",
      "\n",
      "    .. _CELLxGENE Discover:\n",
      "        https://cellxgene.cziscience.com/\n",
      "\n",
      "    .. _cellxgene_census repo:\n",
      "        https://github.com/chanzuckerberg/cellxgene-census/\n",
      "\n",
      "    .. _tiledbsoma repo:\n",
      "        https://github.com/single-cell-data/TileDB-SOMA\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    _experiment\n",
      "    _get_anndata\n",
      "    _open\n",
      "    _presence_matrix\n",
      "    _release_directory\n",
      "    _testing (package)\n",
      "    _util\n",
      "    experimental (package)\n",
      "\n",
      "FUNCTIONS\n",
      "    download_source_h5ad(\n",
      "        dataset_id: str,\n",
      "        to_path: str,\n",
      "        *,\n",
      "        census_version: str = 'stable',\n",
      "        progress_bar: bool = True\n",
      "    ) -> None\n",
      "        Download the source H5AD dataset, for the given `dataset_id`, to the user-specified\n",
      "        file name.\n",
      "\n",
      "        Args:\n",
      "            dataset_id\n",
      "                Fetch the source (original) H5AD associated with this ``dataset_id``.\n",
      "            to_path:\n",
      "                The file name where the downloaded H5AD will be written. Must not already exist.\n",
      "            census_version:\n",
      "                The census version name. Defaults to ``\"stable\"``.\n",
      "            progress_bar:\n",
      "                Whether to display a progress bar. Defaults to ``True``.\n",
      "\n",
      "        Raises:\n",
      "            ValueError: if the path already exists (i.e., will not overwrite an existing file), or is not a file.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        See Also:\n",
      "            :func:`get_source_h5ad_uri`: Look up the location of the source H5AD.\n",
      "\n",
      "        Examples:\n",
      "            >>> download_source_h5ad(\"8e47ed12-c658-4252-b126-381df8d52a3d\", to_path=\"/tmp/data.h5ad\")\n",
      "\n",
      "    get_anndata(\n",
      "        census: tiledbsoma._collection.Collection,\n",
      "        organism: str,\n",
      "        measurement_name: str = 'RNA',\n",
      "        X_name: str = 'raw',\n",
      "        X_layers: collections.abc.Sequence[str] | None = (),\n",
      "        obsm_layers: collections.abc.Sequence[str] | None = (),\n",
      "        obsp_layers: collections.abc.Sequence[str] | None = (),\n",
      "        varm_layers: collections.abc.Sequence[str] | None = (),\n",
      "        varp_layers: collections.abc.Sequence[str] | None = (),\n",
      "        obs_value_filter: str | None = None,\n",
      "        obs_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = None,\n",
      "        var_value_filter: str | None = None,\n",
      "        var_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = None,\n",
      "        column_names: somacore.query.query.AxisColumnNames | None = None,\n",
      "        obs_embeddings: collections.abc.Sequence[str] | None = (),\n",
      "        var_embeddings: collections.abc.Sequence[str] | None = (),\n",
      "        obs_column_names: collections.abc.Sequence[str] | None = None,\n",
      "        var_column_names: collections.abc.Sequence[str] | None = None,\n",
      "        modality: str = 'census_data'\n",
      "    ) -> anndata._core.anndata.AnnData\n",
      "        Convenience wrapper around :class:`tiledbsoma.Experiment` query, to build and execute a query,\n",
      "        and return it as an :class:`anndata.AnnData` object.\n",
      "\n",
      "        Args:\n",
      "            census:\n",
      "                The census object, usually returned by :func:`open_soma`.\n",
      "            organism:\n",
      "                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``.\n",
      "            measurement_name:\n",
      "                The measurement object to query. Defaults to ``\"RNA\"``.\n",
      "            X_name:\n",
      "                The ``X`` layer to query. Defaults to ``\"raw\"``.\n",
      "            X_layers:\n",
      "                Additional layers to add to :attr:`anndata.AnnData.layers`.\n",
      "            obs_value_filter:\n",
      "                Value filter for the ``obs`` metadata. Value is a filter query written in the\n",
      "                SOMA ``value_filter`` syntax.\n",
      "            obs_coords:\n",
      "                Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "            var_value_filter:\n",
      "                Value filter for the ``var`` metadata. Value is a filter query written in the\n",
      "                SOMA ``value_filter`` syntax.\n",
      "            var_coords:\n",
      "                Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "            obsm_layers:\n",
      "                Additional obsm layers to read and return in the ``obsm`` slot.\n",
      "            obsp_layers:\n",
      "                Additional obsp layers to read and return in the ``obsp`` slot.\n",
      "            varm_layers:\n",
      "                Additional varm layers to read and return in the ``varm`` slot.\n",
      "            varp_layers:\n",
      "                Additional varp layers to read and return in the ``varp`` slot.\n",
      "            obs_embeddings:\n",
      "                Additional embeddings to be returned as part of the ``obsm`` slot.\n",
      "                Use :func:`get_all_available_embeddings` to retrieve available embeddings\n",
      "                for this Census version and organism.\n",
      "            var_embeddings:\n",
      "                Additional embeddings to be returned as part of the ``varm`` slot.\n",
      "                Use :func:`get_all_available_embeddings` to retrieve available embeddings\n",
      "                for this Census version and organism.\n",
      "            obs_column_names:\n",
      "                Columns to fetch for ``obs`` dataframe.\n",
      "            var_column_names:\n",
      "                Columns to fetch for ``var`` dataframe.\n",
      "            modality:\n",
      "                Which modality to query, can be one of ``\"census_data\"`` or ``\"census_spatial_sequencing\"``.\n",
      "                Defaults to ``\"census_data\"``.\n",
      "\n",
      "        Returns:\n",
      "            An :class:`anndata.AnnData` object containing the census slice.\n",
      "\n",
      "        Lifecycle:\n",
      "            experimental\n",
      "\n",
      "        Examples:\n",
      "            >>> get_anndata(census, \"Mus musculus\", obs_value_filter=\"tissue_general in ['brain', 'lung']\")\n",
      "\n",
      "            >>> get_anndata(census, \"Homo sapiens\", obs_column_names=[\"tissue\"])\n",
      "\n",
      "            >>> get_anndata(census, \"Homo sapiens\", obs_coords=slice(0, 1000))\n",
      "\n",
      "    get_census_mirror_directory() -> dict[str, cellxgene_census._release_directory.CensusMirror]\n",
      "        Get the directory of Census mirrors currently available.\n",
      "\n",
      "        Returns:\n",
      "            A dictionary that contains mirror names and their corresponding info,\n",
      "            like the provider and the region.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "    get_census_version_description(census_version: str) -> cellxgene_census._release_directory.CensusVersionDescription\n",
      "        Get release description for given Census version, from the Census release directory.\n",
      "\n",
      "        Args:\n",
      "            census_version:\n",
      "                The census version name.\n",
      "\n",
      "        Returns:\n",
      "            ``CensusVersionDescription`` - a dictionary containing a description of the release.\n",
      "\n",
      "        Raises:\n",
      "            ValueError: if unknown ``census_version`` value.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        See Also:\n",
      "            :func:`get_census_version_directory`: returns the entire directory as a dict.\n",
      "\n",
      "        Examples:\n",
      "            >>> cellxgene_census.get_census_version_description(\"latest\")\n",
      "            {'release_date': None,\n",
      "            'release_build': '2022-12-01',\n",
      "            'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n",
      "            's3_region': 'us-west-2'},\n",
      "            'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n",
      "            's3_region': 'us-west-2'}}\n",
      "\n",
      "    get_census_version_directory(\n",
      "        *,\n",
      "        lts: bool | None = None,\n",
      "        retracted: bool | None = False\n",
      "    ) -> dict[str, cellxgene_census._release_directory.CensusVersionDescription]\n",
      "        Get the directory of Census versions currently available, optionally filtering by specified\n",
      "        flags. If a filtering flag is not specified, Census versions will not be filtered by that flag.\n",
      "        Defaults to including both \"long-term stable\" (LTS) and weekly Census versions, and excluding\n",
      "        retracted versions.\n",
      "\n",
      "        Args:\n",
      "            lts:\n",
      "                A filtering flag to either include or exclude long-term stable releases in the result.\n",
      "                If ``None``, no filtering is performed based on this flag. Defaults to ``None``, which\n",
      "                includes both LTS and non-LTS (weekly) versions.\n",
      "            retracted:\n",
      "                A filtering flag to either include or exclude retracted releases in the result. If\n",
      "                ``None``, no filtering is performed based on this flag. Defaults to ``False``, which\n",
      "                excludes retracted releases in the result.\n",
      "\n",
      "        Returns:\n",
      "            A dictionary that contains Census version names and their corresponding descriptions. Census\n",
      "            versions are always named by their release date (``YYYY-MM-DD``) but may also have aliases.\n",
      "            If an alias is specified, the Census version will appear multiple times in the dictionary,\n",
      "            once under it's release date name, and again for each alias. Aliases may be: ``\"stable\"``,\n",
      "            ``\"latest\"``, or ``\"V#\"``. The ``\"stable\"`` alias is used for the most recent LTS release,\n",
      "            the ``\"latest\"`` alias is used for the most recent weekly release, and the ``\"V#\"`` aliases\n",
      "            are used to identify LTS releases by a sequentially incrementing version number.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        See Also:\n",
      "            :func:`get_census_version_description`: get description by ``census_version``.\n",
      "\n",
      "        Examples:\n",
      "            Get all LTS and weekly versions, but exclude retracted LTS versions:\n",
      "\n",
      "            >>> cellxgene_census.get_census_version_directory()\n",
      "                {\n",
      "                    'stable': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    },\n",
      "                    'latest': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-12-01',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    },\n",
      "                    'V2': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    },\n",
      "                    '2022-12-01': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-12-01',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': False, 'retracted': False}\n",
      "                    },\n",
      "                    '2022-11-29': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    }\n",
      "                }\n",
      "\n",
      "            Get only LTS versions that are not retracted:\n",
      "\n",
      "            >>> cellxgene_census.get_census_version_directory(lts=True)\n",
      "                {\n",
      "                    'stable': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    },\n",
      "                    'V2': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    },\n",
      "                    '2022-11-29': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-11-29',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-11-29/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': False}\n",
      "                    }\n",
      "                }\n",
      "\n",
      "            Get only retracted releases:\n",
      "\n",
      "            >>> cellxgene_census.get_census_version_directory(retracted=True)\n",
      "                {\n",
      "                    'V1': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-10-15',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': True},\n",
      "                        'retraction': {\n",
      "                            'date': '2022-10-30',\n",
      "                            'reason': 'mistakes happen',\n",
      "                            'info_url': 'http://cellxgene.com/census/errata/v1',\n",
      "                            'replaced_by': 'V2'\n",
      "                        },\n",
      "                    },\n",
      "                    '2022-10-15': {\n",
      "                        'release_date': None,\n",
      "                        'release_build': '2022-10-15',\n",
      "                        'soma': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/soma/',\n",
      "                                 's3_region': 'us-west-2'},\n",
      "                        'h5ads': {'uri': 's3://cellxgene-data-public/cell-census/2022-10-15/h5ads/',\n",
      "                                  's3_region': 'us-west-2'},\n",
      "                        'flags': {'lts': True, 'retracted': True},\n",
      "                        'retraction': {\n",
      "                            'date': '2022-10-30',\n",
      "                            'reason': 'mistakes happen',\n",
      "                            'info_url': 'http://cellxgene.com/census/errata/v1',\n",
      "                            'replaced_by': 'V2'\n",
      "                        }\n",
      "                    }\n",
      "                }\n",
      "\n",
      "    get_default_soma_context(tiledb_config: dict[str, typing.Any] | None = None) -> tiledbsoma.options._soma_tiledb_context.SOMATileDBContext\n",
      "        Return a :class:`tiledbsoma.SOMATileDBContext` with sensible defaults that can be further customized by the\n",
      "        user. The customized context can then be passed to :func:`cellxgene_census.open_soma` with the ``context``\n",
      "        argument or to :meth:`somacore.SOMAObject.open` with the ``context`` argument, such as\n",
      "        :meth:`tiledbsoma.Experiment.open`. Use the :meth:`tiledbsoma.SOMATileDBContext.replace` method on the returned\n",
      "        object to customize its settings further.\n",
      "\n",
      "        Args:\n",
      "            tiledb_config:\n",
      "                A dictionary of TileDB configuration parameters. If specified, the parameters will override the\n",
      "                defaults. If not specified, the default configuration will be returned.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`tiledbsoma.SOMATileDBContext` object with sensible defaults.\n",
      "\n",
      "        Examples:\n",
      "            To reduce the amount of memory used by TileDB-SOMA I/O operations:\n",
      "\n",
      "            .. highlight:: python\n",
      "            .. code-block:: python\n",
      "\n",
      "                ctx = cellxgene_census.get_default_soma_context(\n",
      "                    tiledb_config={\"py.init_buffer_bytes\": 128 * 1024**2,\n",
      "                                   \"soma.init_buffer_bytes\": 128 * 1024**2})\n",
      "                c = census.open_soma(uri=\"s3://my-private-bucket/census/soma\", context=ctx)\n",
      "\n",
      "            To access a copy of the Census located in a private bucket that is located in a different S3 region, use:\n",
      "\n",
      "            .. highlight:: python\n",
      "            .. code-block:: python\n",
      "\n",
      "                ctx = cellxgene_census.get_default_soma_context(\n",
      "                    tiledb_config={\"vfs.s3.no_sign_request\": \"false\",\n",
      "                                   \"vfs.s3.region\": \"us-east-1\"})\n",
      "                c = census.open_soma(uri=\"s3://my-private-bucket/census/soma\", context=ctx)\n",
      "\n",
      "        Lifecycle:\n",
      "            experimental\n",
      "\n",
      "    get_obs(\n",
      "        census: tiledbsoma._collection.Collection,\n",
      "        organism: str,\n",
      "        *,\n",
      "        value_filter: str | None = None,\n",
      "        coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = slice(None, None, None),\n",
      "        column_names: collections.abc.Sequence[str] | None = None,\n",
      "        modality: str = 'census_data'\n",
      "    ) -> pandas.core.frame.DataFrame\n",
      "        Get the observation metadata for a query on the census.\n",
      "\n",
      "        Args:\n",
      "            census:\n",
      "                The census object, usually returned by :func:`open_soma`.\n",
      "            organism:\n",
      "                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``\n",
      "            value_filter:\n",
      "                Value filter for the ``obs`` metadata. Value is a filter query written in the\n",
      "                SOMA ``value_filter`` syntax.\n",
      "            coords:\n",
      "                Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "            column_names:\n",
      "                Columns to fetch.\n",
      "            modality\n",
      "                Which modality to query, can be one of ``\"census_data\"`` or ``\"census_spatial_sequencing\"``.\n",
      "                Defaults to ``\"census_data\"``.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`pandas.DataFrame` object containing metadata for the queried slice.\n",
      "\n",
      "    get_presence_matrix(\n",
      "        census: tiledbsoma._collection.Collection,\n",
      "        organism: str,\n",
      "        measurement_name: str = 'RNA',\n",
      "        modality: str = 'census_data'\n",
      "    ) -> scipy.sparse._csr.csr_matrix\n",
      "        Read the feature dataset presence matrix and return as a :class:`scipy.sparse.csr_array`. The\n",
      "        returned sparse matrix is indexed on the first dimension by the dataset ``soma_joinid`` values,\n",
      "        and on the second dimension by the ``var`` :class:`pandas.DataFrame` ``soma_joinid`` values.\n",
      "\n",
      "        Args:\n",
      "            census:\n",
      "                The census from which to read the presence matrix.\n",
      "            organism:\n",
      "                The organism to query, usually one of ``\"Homo sapiens\"`` or ``\"Mus musculus\"``.\n",
      "            measurement_name:\n",
      "                The measurement object to query. Deafults to ``\"RNA\"``.\n",
      "            modality:\n",
      "                Which modality to query, can be one of ``\"census_data\"`` or ``\"census_spatial_sequencing\"``.\n",
      "                Defaults to ``\"census_data\"``.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`scipy.sparse.csr_array` object containing the presence matrix.\n",
      "\n",
      "        Raises:\n",
      "            ValueError: if the organism cannot be found.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        Examples:\n",
      "            >>> get_presence_matrix(census, \"Homo sapiens\", \"RNA\")\n",
      "            <321x60554 sparse array of type '<class 'numpy.uint8'>'\n",
      "            with 6441269 stored elements in Compressed Sparse Row format>\n",
      "\n",
      "    get_source_h5ad_uri(dataset_id: str, *, census_version: str = 'stable') -> cellxgene_census._release_directory.CensusLocator\n",
      "        Open the named version of the census, and return the URI for the ``dataset_id``. This\n",
      "        does not guarantee that the H5AD exists or is accessible to the user.\n",
      "\n",
      "        Args:\n",
      "            dataset_id:\n",
      "                The ``dataset_id`` of interest.\n",
      "            census_version:\n",
      "                The census version. Defaults to ``\"stable\"``.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`cellxgene_census._release_directory.CensusLocator` object that contains the URI and optional S3 region\n",
      "            for the source H5AD.\n",
      "\n",
      "        Raises:\n",
      "            KeyError: if either ``dataset_id`` or ``census_version`` do not exist.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        Examples:\n",
      "            >>> cellxgene_census.get_source_h5ad_uri(\"cb5efdb0-f91c-4cbd-9ad4-9d4fa41c572d\")\n",
      "            {'uri': 's3://cellxgene-data-public/cell-census/2022-12-01/h5ads/cb5efdb0-f91c-4cbd-9ad4-9d4fa41c572d.h5ad',\n",
      "            's3_region': 'us-west-2'}\n",
      "\n",
      "    get_var(\n",
      "        census: tiledbsoma._collection.Collection,\n",
      "        organism: str,\n",
      "        *,\n",
      "        value_filter: str | None = None,\n",
      "        coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = slice(None, None, None),\n",
      "        column_names: collections.abc.Sequence[str] | None = None,\n",
      "        modality: str = 'census_data'\n",
      "    ) -> pandas.core.frame.DataFrame\n",
      "        Get the variable metadata for a query on the census.\n",
      "\n",
      "        Args:\n",
      "            census:\n",
      "                The census object, usually returned by :func:`open_soma`.\n",
      "            organism:\n",
      "                The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``\n",
      "            value_filter:\n",
      "                Value filter for the ``var`` metadata. Value is a filter query written in the\n",
      "                SOMA ``value_filter`` syntax.\n",
      "            coords:\n",
      "                Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "                May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "            column_names:\n",
      "                Columns to fetch.\n",
      "            modality:\n",
      "                Which modality to query, can be one of ``\"census_data\"`` or ``\"census_spatial_sequencing\"``.\n",
      "                Defaults to ``\"census_data\"``.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`pandas.DataFrame` object containing metadata for the queried slice.\n",
      "\n",
      "    open_soma(\n",
      "        *,\n",
      "        census_version: str | None = 'stable',\n",
      "        mirror: str | None = None,\n",
      "        uri: str | None = None,\n",
      "        tiledb_config: dict[str, typing.Any] | None = None,\n",
      "        context: tiledbsoma.options._soma_tiledb_context.SOMATileDBContext | None = None\n",
      "    ) -> tiledbsoma._collection.Collection\n",
      "        Open the Census by version or URI.\n",
      "\n",
      "        Args:\n",
      "            census_version:\n",
      "                The version of the Census, e.g. ``\"latest\"`` or ``\"stable\"``. Defaults to ``\"stable\"``.\n",
      "            mirror:\n",
      "                The mirror used to retrieve the Census. If not specified, a suitable mirror\n",
      "                will be chosen automatically.\n",
      "            uri:\n",
      "                The URI containing the Census SOMA objects. If specified, will take precedence\n",
      "                over ``census_version`` parameter.\n",
      "            tiledb_config:\n",
      "                A dictionary of TileDB configuration parameters that will be used to open the SOMA object. Optional,\n",
      "                defaults to ``None``. If specified, the parameters will override the default settings specified by\n",
      "                ``get_default_soma_context().tiledb_config``. Only one of the ``tiledb_config`` and ``context`` params\n",
      "                can be specified.\n",
      "            context:\n",
      "                A custom :class:`tiledbsoma.SOMATileDBContext` that will be used to open the SOMA object.\n",
      "                Optional, defaults to ``None``. Only one of the ``tiledb_config`` and ``context`` params can be specified.\n",
      "\n",
      "        Returns:\n",
      "            A :class:`tiledbsoma.Collection` object containing the top-level census.\n",
      "            It can be used as a context manager, which will automatically close upon exit.\n",
      "\n",
      "        Raises:\n",
      "            ValueError: if the census cannot be found, the URI cannot be opened, neither a URI\n",
      "                or a version are specified, or an invalid mirror is provided.\n",
      "\n",
      "        See Also:\n",
      "            - :func:`get_source_h5ad_uri`: Look up the location of the source H5AD.\n",
      "            - :func:`get_default_soma_context`: Get a default SOMA context that can be updated with custom settings and used\n",
      "              as the ``context`` argument.\n",
      "\n",
      "        Lifecycle:\n",
      "            maturing\n",
      "\n",
      "        Examples:\n",
      "            Open the default Census version, using a context manager which will automatically\n",
      "            close the Census upon exit of the context.\n",
      "\n",
      "            >>> with cellxgene_census.open_soma() as census:\n",
      "                    ...\n",
      "\n",
      "            Open and close:\n",
      "\n",
      "            >>> census = cellxgene_census.open_soma()\n",
      "                ...\n",
      "                census.close()\n",
      "\n",
      "            Open a specific Census by version:\n",
      "\n",
      "            >>> with cellxgene_census.open_soma(\"2022-12-31\") as census:\n",
      "                    ...\n",
      "\n",
      "            Open a Census by S3 URI, rather than by version.\n",
      "\n",
      "            >>> with cellxgene_census.open_soma(uri=\"s3://bucket/path\") as census:\n",
      "                    ...\n",
      "\n",
      "            Open a Census by path (file:// URI), rather than by version.\n",
      "\n",
      "            >>> with cellxgene_census.open_soma(uri=\"/tmp/census\") as census:\n",
      "                    ...\n",
      "\n",
      "            Open a Census using a mirror.\n",
      "\n",
      "            >>> with cellxgene_census.open_soma(mirror=\"s3-us-west-2\") as census:\n",
      "                    ...\n",
      "\n",
      "            Open a Census with a custom TileDB configuration setting.\n",
      "\n",
      "            >>> with cellxgene_census.open_soma(tiledb_config={\"py.init_buffer_bytes\": 128 * 1024**2}) as census:\n",
      "                    ...\n",
      "\n",
      "DATA\n",
      "    __all__ = ['download_source_h5ad', 'get_anndata', 'get_obs', 'get_var'...\n",
      "\n",
      "VERSION\n",
      "    1.17.0\n",
      "\n",
      "FILE\n",
      "    /Users/hweej/Projects/Sandbox/cancer-cell-portal-utils/.venv/lib/python3.13/site-packages/cellxgene_census/__init__.py\n",
      "\n",
      "\n",
      "Help on function get_anndata in module cellxgene_census._get_anndata:\n",
      "\n",
      "get_anndata(\n",
      "    census: tiledbsoma._collection.Collection,\n",
      "    organism: str,\n",
      "    measurement_name: str = 'RNA',\n",
      "    X_name: str = 'raw',\n",
      "    X_layers: collections.abc.Sequence[str] | None = (),\n",
      "    obsm_layers: collections.abc.Sequence[str] | None = (),\n",
      "    obsp_layers: collections.abc.Sequence[str] | None = (),\n",
      "    varm_layers: collections.abc.Sequence[str] | None = (),\n",
      "    varp_layers: collections.abc.Sequence[str] | None = (),\n",
      "    obs_value_filter: str | None = None,\n",
      "    obs_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = None,\n",
      "    var_value_filter: str | None = None,\n",
      "    var_coords: Union[NoneType, bytes, somacore.types.Slice[bytes], Sequence[bytes], float, somacore.types.Slice[float], Sequence[float], int, somacore.types.Slice[int], Sequence[int], slice, somacore.types.Slice[slice], Sequence[slice], str, somacore.types.Slice[str], Sequence[str], numpy.datetime64, somacore.types.Slice[numpy.datetime64], Sequence[numpy.datetime64], pyarrow.lib.TimestampType, somacore.types.Slice[pyarrow.lib.TimestampType], Sequence[pyarrow.lib.TimestampType], pyarrow.lib.Array, pyarrow.lib.ChunkedArray, numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.integer]], numpy.ndarray[tuple[int, ...], numpy.dtype[numpy.datetime64]]] = None,\n",
      "    column_names: somacore.query.query.AxisColumnNames | None = None,\n",
      "    obs_embeddings: collections.abc.Sequence[str] | None = (),\n",
      "    var_embeddings: collections.abc.Sequence[str] | None = (),\n",
      "    obs_column_names: collections.abc.Sequence[str] | None = None,\n",
      "    var_column_names: collections.abc.Sequence[str] | None = None,\n",
      "    modality: str = 'census_data'\n",
      ") -> anndata._core.anndata.AnnData\n",
      "    Convenience wrapper around :class:`tiledbsoma.Experiment` query, to build and execute a query,\n",
      "    and return it as an :class:`anndata.AnnData` object.\n",
      "\n",
      "    Args:\n",
      "        census:\n",
      "            The census object, usually returned by :func:`open_soma`.\n",
      "        organism:\n",
      "            The organism to query, usually one of ``\"Homo sapiens`` or ``\"Mus musculus\"``.\n",
      "        measurement_name:\n",
      "            The measurement object to query. Defaults to ``\"RNA\"``.\n",
      "        X_name:\n",
      "            The ``X`` layer to query. Defaults to ``\"raw\"``.\n",
      "        X_layers:\n",
      "            Additional layers to add to :attr:`anndata.AnnData.layers`.\n",
      "        obs_value_filter:\n",
      "            Value filter for the ``obs`` metadata. Value is a filter query written in the\n",
      "            SOMA ``value_filter`` syntax.\n",
      "        obs_coords:\n",
      "            Coordinates for the ``obs`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "            May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "        var_value_filter:\n",
      "            Value filter for the ``var`` metadata. Value is a filter query written in the\n",
      "            SOMA ``value_filter`` syntax.\n",
      "        var_coords:\n",
      "            Coordinates for the ``var`` axis, which is indexed by the ``soma_joinid`` value.\n",
      "            May be an ``int``, a list of ``int``, or a slice. The default, ``None``, selects all.\n",
      "        obsm_layers:\n",
      "            Additional obsm layers to read and return in the ``obsm`` slot.\n",
      "        obsp_layers:\n",
      "            Additional obsp layers to read and return in the ``obsp`` slot.\n",
      "        varm_layers:\n",
      "            Additional varm layers to read and return in the ``varm`` slot.\n",
      "        varp_layers:\n",
      "            Additional varp layers to read and return in the ``varp`` slot.\n",
      "        obs_embeddings:\n",
      "            Additional embeddings to be returned as part of the ``obsm`` slot.\n",
      "            Use :func:`get_all_available_embeddings` to retrieve available embeddings\n",
      "            for this Census version and organism.\n",
      "        var_embeddings:\n",
      "            Additional embeddings to be returned as part of the ``varm`` slot.\n",
      "            Use :func:`get_all_available_embeddings` to retrieve available embeddings\n",
      "            for this Census version and organism.\n",
      "        obs_column_names:\n",
      "            Columns to fetch for ``obs`` dataframe.\n",
      "        var_column_names:\n",
      "            Columns to fetch for ``var`` dataframe.\n",
      "        modality:\n",
      "            Which modality to query, can be one of ``\"census_data\"`` or ``\"census_spatial_sequencing\"``.\n",
      "            Defaults to ``\"census_data\"``.\n",
      "\n",
      "    Returns:\n",
      "        An :class:`anndata.AnnData` object containing the census slice.\n",
      "\n",
      "    Lifecycle:\n",
      "        experimental\n",
      "\n",
      "    Examples:\n",
      "        >>> get_anndata(census, \"Mus musculus\", obs_value_filter=\"tissue_general in ['brain', 'lung']\")\n",
      "\n",
      "        >>> get_anndata(census, \"Homo sapiens\", obs_column_names=[\"tissue\"])\n",
      "\n",
      "        >>> get_anndata(census, \"Homo sapiens\", obs_coords=slice(0, 1000))\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import cellxgene_census\n",
    "\n",
    "help(cellxgene_census)\n",
    "help(cellxgene_census.get_anndata)\n",
    "# etc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fe559aaf-bd45-4079-909b-254f7198bfd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "            assay        cell_type                          tissue  \\\n",
      "0       10x 5' v1           neuron                            lung   \n",
      "1       10x 5' v1           neuron                            lung   \n",
      "2       10x 5' v1           neuron                            lung   \n",
      "3       10x 5' v1           neuron                            lung   \n",
      "4       10x 5' v1           neuron                            lung   \n",
      "...           ...              ...                             ...   \n",
      "796421  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n",
      "796422  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n",
      "796423  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n",
      "796424  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n",
      "796425  10x 3' v3  microglial cell  dorsolateral prefrontal cortex   \n",
      "\n",
      "       tissue_general suspension_type   disease     sex  \n",
      "0                lung            cell    normal  female  \n",
      "1                lung            cell    normal  female  \n",
      "2                lung            cell    normal  female  \n",
      "3                lung            cell    normal  female  \n",
      "4                lung            cell    normal  female  \n",
      "...               ...             ...       ...     ...  \n",
      "796421          brain         nucleus    normal  female  \n",
      "796422          brain         nucleus    normal  female  \n",
      "796423          brain         nucleus  dementia  female  \n",
      "796424          brain         nucleus  dementia  female  \n",
      "796425          brain         nucleus  dementia  female  \n",
      "\n",
      "[796426 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "# Querying a slice of cell metadata\n",
    "with cellxgene_census.open_soma(census_version=\"2025-01-30\") as census:\n",
    "    # Reads SOMADataFrame as a slice\n",
    "    cell_metadata = census[\"census_data\"][\"homo_sapiens\"].obs.read(\n",
    "        value_filter = \"sex == 'female' and cell_type in ['microglial cell', 'neuron']\",\n",
    "        column_names = [\"assay\", \"cell_type\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]\n",
    "    )\n",
    "\n",
    "    # Concatenates results to pyarrow.Table\n",
    "    cell_metadata = cell_metadata.concat()\n",
    "\n",
    "    # Converts to pandas.DataFrame\n",
    "    cell_metadata = cell_metadata.to_pandas()\n",
    "\n",
    "    print(cell_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "651bebf4-5732-4edf-9f1a-0060475ff83e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Hangs and never completes, memory continue to climbs\n",
    "\"\"\"\n",
    "# Obtaining a slice as AnnData\n",
    "with cellxgene_census.open_soma(census_version=\"2025-01-30\") as census:\n",
    "    adata = cellxgene_census.get_anndata(\n",
    "        census = census,\n",
    "        organism = \"Homo sapiens\",\n",
    "        var_value_filter = \"feature_id in ['ENSG00000161798', 'ENSG00000188229']\",\n",
    "        obs_value_filter = \"sex == 'female' and cell_type in ['microglial cell', 'neuron']\",\n",
    "        obs_column_names = {\"obs\": [\"assay\", \"cell_type\", \"tissue\", \"tissue_general\", \"suspension_type\", \"disease\"]}\n",
    "    )\n",
    "\n",
    "    print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "973886aa-cf6d-4f22-b9cf-3bef10ccb8c1",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mKeyboardInterrupt\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[8]\u001b[39m\u001b[32m, line 19\u001b[39m\n\u001b[32m      8\u001b[39m query = human.axis_query(\n\u001b[32m      9\u001b[39m     measurement_name = \u001b[33m\"\u001b[39m\u001b[33mRNA\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m     10\u001b[39m     obs_query = tiledbsoma.AxisQuery(\n\u001b[32m     11\u001b[39m         value_filter = \u001b[33m\"\u001b[39m\u001b[33mtissue == \u001b[39m\u001b[33m'\u001b[39m\u001b[33mbrain\u001b[39m\u001b[33m'\u001b[39m\u001b[33m and sex == \u001b[39m\u001b[33m'\u001b[39m\u001b[33mmale\u001b[39m\u001b[33m'\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m     12\u001b[39m     )\n\u001b[32m     13\u001b[39m )\n\u001b[32m     14\u001b[39m \u001b[38;5;250m\u001b[39m\u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m     15\u001b[39m \u001b[33;03mNow we can iterate over the matrix count, as we as the gene metadata.\u001b[39;00m\n\u001b[32m     16\u001b[39m \u001b[33;03mFor example, to iterate over the matrix count, we can get an iterator\u001b[39;00m\n\u001b[32m     17\u001b[39m \u001b[33;03mand perform operations for each iteration.\u001b[39;00m\n\u001b[32m     18\u001b[39m \u001b[33;03m\"\"\"\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m19\u001b[39m iterator = \u001b[43mquery\u001b[49m\u001b[43m.\u001b[49m\u001b[43mX\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m\"\u001b[39;49m\u001b[33;43mraw\u001b[39;49m\u001b[33;43m\"\u001b[39;49m\u001b[43m)\u001b[49m.tables()\n\u001b[32m     21\u001b[39m \u001b[38;5;66;03m# Get an iterative slice as pyarrow.Table\u001b[39;00m\n\u001b[32m     22\u001b[39m raw_slice = \u001b[38;5;28mnext\u001b[39m (iterator)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Projects/Sandbox/cancer-cell-portal-utils/.venv/lib/python3.13/site-packages/tiledbsoma/_query.py:304\u001b[39m, in \u001b[36mExperimentAxisQuery.X\u001b[39m\u001b[34m(self, layer_name, batch_size, partitions, result_order, platform_config)\u001b[39m\n\u001b[32m    301\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(x_layer, SparseNDArray):\n\u001b[32m    302\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\u001b[33m\"\u001b[39m\u001b[33mX layers may only be sparse arrays\u001b[39m\u001b[33m\"\u001b[39m)\n\u001b[32m--> \u001b[39m\u001b[32m304\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_joinids\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpreload\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_threadpool\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    305\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m x_layer.read(\n\u001b[32m    306\u001b[39m     (\u001b[38;5;28mself\u001b[39m._joinids.obs, \u001b[38;5;28mself\u001b[39m._joinids.var),\n\u001b[32m    307\u001b[39m     batch_size=batch_size,\n\u001b[32m   (...)\u001b[39m\u001b[32m    310\u001b[39m     platform_config=platform_config,\n\u001b[32m    311\u001b[39m )\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/Projects/Sandbox/cancer-cell-portal-utils/.venv/lib/python3.13/site-packages/tiledbsoma/_query.py:751\u001b[39m, in \u001b[36mJoinIDCache.preload\u001b[39m\u001b[34m(self, pool)\u001b[39m\n\u001b[32m    749\u001b[39m var_ft = pool.submit(\u001b[38;5;28;01mlambda\u001b[39;00m: \u001b[38;5;28mself\u001b[39m.var)\n\u001b[32m    750\u001b[39m \u001b[38;5;66;03m# Wait for them and raise in case of error.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m751\u001b[39m \u001b[43mobs_ft\u001b[49m\u001b[43m.\u001b[49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    752\u001b[39m var_ft.result()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.13.0/lib/python3.13/concurrent/futures/_base.py:451\u001b[39m, in \u001b[36mFuture.result\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    448\u001b[39m \u001b[38;5;28;01melif\u001b[39;00m \u001b[38;5;28mself\u001b[39m._state == FINISHED:\n\u001b[32m    449\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m.__get_result()\n\u001b[32m--> \u001b[39m\u001b[32m451\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_condition\u001b[49m\u001b[43m.\u001b[49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    453\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m._state \u001b[38;5;129;01min\u001b[39;00m [CANCELLED, CANCELLED_AND_NOTIFIED]:\n\u001b[32m    454\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m CancelledError()\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.13.0/lib/python3.13/threading.py:359\u001b[39m, in \u001b[36mCondition.wait\u001b[39m\u001b[34m(self, timeout)\u001b[39m\n\u001b[32m    357\u001b[39m \u001b[38;5;28;01mtry\u001b[39;00m:    \u001b[38;5;66;03m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[39;00m\n\u001b[32m    358\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m--> \u001b[39m\u001b[32m359\u001b[39m         \u001b[43mwaiter\u001b[49m\u001b[43m.\u001b[49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    360\u001b[39m         gotit = \u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[32m    361\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n",
      "\u001b[31mKeyboardInterrupt\u001b[39m: "
     ]
    }
   ],
   "source": [
    "# Memory-efficient queries\n",
    "\n",
    "import cellxgene_census\n",
    "import tiledbsoma\n",
    "\n",
    "with cellxgene_census.open_soma(census_version=\"2025-01-30\") as census:\n",
    "    human = census[\"census_data\"][\"homo_sapiens\"]\n",
    "    query = human.axis_query(\n",
    "        measurement_name = \"RNA\",\n",
    "        obs_query = tiledbsoma.AxisQuery(\n",
    "            value_filter = \"tissue == 'brain' and sex == 'male'\"\n",
    "        )\n",
    "    )\n",
    "    \"\"\"\n",
    "    Now we can iterate over the matrix count, as we as the gene metadata.\n",
    "    For example, to iterate over the matrix count, we can get an iterator\n",
    "    and perform operations for each iteration.\n",
    "    \"\"\"\n",
    "    iterator = query.X(\"raw\").tables()\n",
    "\n",
    "    # Get an iterative slice as pyarrow.Table\n",
    "    raw_slice = next (iterator)\n",
    "    \"\"\"\n",
    "    And you can now perform operations of each iteration slice. As with\n",
    "    any Python iterator this logic can be wrapped around a for loop.\n",
    "    \"\"\"\n",
    "    query.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c412bb84-5c91-4311-ae4d-cc6aed040684",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
